### PR TITLE
feat: add nydusify commit command

### PR DESCRIPTION
**Pull request by PerseidMeteor**
## Relevant Issue (if applicable)
[support nydus commit operation](https://github.com/dragonflyoss/nydus/issues/1307)

## Details
_Please describe the details of PullRequest._

## Types of changes

_What types of changes does your PullRequest introduce? Put an `x` in all the boxes that apply:_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Go over all the following points, and put an `x` in all the boxes that apply._

- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: contrib/nydusify/pkg/commiter/util.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package commiter
+
+import (
+	// "context"
```
Comment (PerseidMeteor) [inline contrib/nydusify/pkg/commiter/util.go:4]: forget, remove later

**Comment by imeoer** [comment]
Thanks for the implementation! Please also add a test case in https://github.com/dragonflyoss/nydus/tree/master/smoke/tests (maybe create a `commit_test.go` ?).

**Comment by PerseidMeteor** [comment]
> Thanks for the implementation! Please also add a test case in https://github.com/dragonflyoss/nydus/tree/master/smoke/tests (maybe create a `commit_test.go` ?).

I am sorry for I was too late because of my Spring Festival holiday. Commit smoke test was added.
Have a nice day!

**Comment by davidstack** [comment]
very nice pr

**Comment by Desiki-high** [comment]
Thanks for your pr, please fix the `golang-lint`.

File: smoke/tests/commit_test.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	testImage                  string
+	testCommittedImage         string
+	testContainerName          string
+	testCommittedContainerName string
+	snapshotter                string
+}
+
+func (c *CommitTestSuite) TestCommitContainer() {
+	ctx := tool.DefaultContext(c.t)
+	c.snapshotter = os.Getenv("SNAPSHOTTER")
```
Comment (imeoer) [inline smoke/tests/commit_test.go:25]: We'd better use `scenarios := tool.DescartesIterator{}` to manage parameters. :)

File: smoke/tests/commit_test.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+	if c.snapshotter == "" {
+		c.snapshotter = "nydus"
+	}
+
+	// choose test mode
+	mode := os.Getenv("COMMIT_MODE")
```
Comment (imeoer) [inline smoke/tests/commit_test.go:31]: Ditto, should use `tool.DescartesIterator`.

File: smoke/tests/commit_test.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	// make some modifications in the read-write layer for commit action.
+	modifyCmd := fmt.Sprintf("echo 'This is Nydus commit' > commit && nerdctl cp commit %s:/root/", c.testContainerName)
+	tool.RunWithoutOutput(c.t, modifyCmd)
+
+	// commit container
+	commitCmd := fmt.Sprintf("%s commit --container %s --target %s", ctx.Binary.Nydusify, c.testCommittedContainerName, c.testCommittedImage)
+	tool.RunWithoutOutput(c.t, commitCmd)
+
+	// run committed container
+	_ = tool.RunContainer(c.t, c.testCommittedImage, c.snapshotter, c.testCommittedContainerName)
```
Comment (imeoer) [inline smoke/tests/commit_test.go:71]: Should also do `nerdctl cp %s:/root/commit commit` to check the file content. :)

File: smoke/tests/commit_test.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+	if err != nil {
+		t.Fatalf("can't parsing convert metric file")
+		return 0, 0
+	}
+	if c.snapshotter == "nydus" {
+		c.testImage = target
+		return convertMetirc["TargetImageSize"], convertMetirc["ConversionElapsed"]
+	}
+	c.testImage = source
+	return convertMetirc["SourceImageSize"], 0
```
Comment (imeoer) [inline smoke/tests/commit_test.go:123]: Why we need to consider about metric on this test?

File: smoke/tests/commit_test.go [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+func TestCommit(t *testing.T) {
+	if os.Getenv("COMMIT_TEST") == "" {
```
Comment (imeoer) [inline smoke/tests/commit_test.go:127]: In what scenario should this test be skipped?

File: contrib/nydusify/pkg/utils/constant.go [hunk 7]
```diff
@@ -21,4 +21,7 @@ const (
 	LayerAnnotationNydusReferenceBlobIDs = "containerd.io/snapshot/nydus-reference-blob-ids"
 
 	LayerAnnotationUncompressed = "containerd.io/uncompressed"
+
+	LayerAnnotationNydusCommitBlobs = "containerd.io/snapshot/nydus-commit-blobs"
+	LayerAnnotationNydusBlobIDs     = "containerd.io/snapshot/nydus-blob-ids"
```
Comment (imeoer) [inline contrib/nydusify/pkg/utils/constant.go:26]: The constant seems not to be used.

File: contrib/nydusify/cmd/nydusify.go [hunk 8]
```diff
@@ -1103,6 +1104,94 @@ func main() {
...
+					EnvVars: []string{"WORK_DIR"},
+				},
+				&cli.StringFlag{
+					Name:    "nydus-image",
+					Value:   "nydus-image",
+					Usage:   "Path to the nydus-image binary, default to search in PATH",
+					EnvVars: []string{"NYDUS_IMAGE"},
+				},
+				&cli.StringFlag{
+					Name:    "containerd-address",
```
Comment (imeoer) [inline contrib/nydusify/cmd/nydusify.go:1124]: Let's we follow the option help message of nerdctl:

`containerd address, optionally with "unix://" prefix [$CONTAINERD_ADDRESS] (default "/run/containerd/containerd.sock")`.

File: contrib/nydusify/cmd/nydusify.go [hunk 9]
```diff
@@ -1103,6 +1104,94 @@ func main() {
...
+					DefaultText: "400",
+					Value:       400,
+					Usage:       "The maximum times allowed to be committed",
+					EnvVars:     []string{"MAXIMUM_TIMES"},
+				},
+				&cli.StringSliceFlag{
+					Name:     "with-path",
+					Aliases:  []string{"with-mount-path"},
+					Required: false,
+					Usage:    "The directory that need to be committed",
```
Comment (imeoer) [inline contrib/nydusify/cmd/nydusify.go:1153]: The directory in container that need to be committed

File: contrib/nydusify/pkg/commiter/commiter.go [hunk 10]
```diff
@@ -0,0 +1,671 @@
...
+	}
+	return &Commiter{
+		workDir: workDir,
+		builder: opt.NydusImagePath,
+		manager: cm,
+	}, nil
+}
+
+func (cm *Commiter) Commit(ctx context.Context, opt Opt) error {
+	targetRef, err := AppendNydusSuffix(opt.TargetRef)
```
[the remark this answers is not in the document]
Comment (imeoer) [reply contrib/nydusify/pkg/commiter/commiter.go:75]: Right, should use target reference directly.

[the remark this answers is not in the document]
Comment (imeoer) [reply contrib/nydusify/pkg/commiter/commiter.go:75]: Let's remove these suffix logics.

File: contrib/nydusify/pkg/commiter/nsenter.go [hunk 11]
```diff
@@ -0,0 +1,183 @@
+package commiter
```
Comment (imeoer) [inline contrib/nydusify/pkg/commiter/nsenter.go:1]: Should put the copyright, the codes ported from `https://github.com/Devatoria/go-nsenter`.

**Comment by davidstack** [comment]
@imeoer  which release will contain this pr？ we are waiting for this pr for a long time. thanks

**Comment by imeoer** [comment]
> @imeoer which release will contain this pr？ we are waiting for this pr for a long time. thanks

Thanks for waiting, it should be included in the next release, v2.3.0.

**Comment by imeoer** [comment]
@PerseidMeteor Please also update the docs in https://github.com/dragonflyoss/nydus/blob/master/docs/nydusify.md, better be a workflow like `nydusify convert -> nerdctl run -> nydusify commit -> nydusify run`.

File: contrib/nydusify/cmd/nydusify.go [hunk 12]
```diff
@@ -1103,6 +1104,108 @@ func main() {
...
+						} else {
+							withPaths = append(withPaths, path)
+						}
+					}
+
+					return withPaths, withoutPaths
+				}
+
+				withPaths, withoutPaths := parsePaths(c.StringSlice("with-path"))
+				opt := commiter.Opt{
```
Comment (imeoer) [inline contrib/nydusify/cmd/nydusify.go:1190]: Sorry for the misleading, it should be `committer`.

File: contrib/nydusify/cmd/nydusify.go [hunk 13]
```diff
@@ -1103,6 +1104,108 @@ func main() {
...
+					ContainerdAddress: c.String("containerd-address"),
+					ContainerID:       c.String("container"),
+					TargetRef:         c.String("target"),
+					SourceInsecure:    c.Bool("source-insecure"),
+					TargetInsecure:    c.Bool("target-insecure"),
+					MaximumTimes:      c.Int("maximum-times"),
+					WithPaths:         withPaths,
+					WithoutPaths:      withoutPaths,
+				}
+				cm, err := commiter.NewCommiter(opt)
```
Comment (imeoer) [inline contrib/nydusify/cmd/nydusify.go:1202]: Ditto.

File: contrib/nydusify/cmd/nydusify.go [hunk 14]
```diff
@@ -1103,6 +1104,108 @@ func main() {
...
+				if err != nil {
+					return errors.Wrap(err, "create commiter")
+				}
+				return cm.Commit(context.Background(), opt)
```
Comment (imeoer) [inline contrib/nydusify/cmd/nydusify.go:1206]: `context.Background()` -> `c.Context()`.

File: contrib/nydusify/pkg/commiter/diff/diff.go [hunk 15]
```diff
@@ -0,0 +1,113 @@
+// Ported from buildkit project, copyright The buildkit Authors.
+// https://github.com/moby/buildkit
```
Comment (imeoer) [inline contrib/nydusify/pkg/commiter/diff/diff.go:2]: Leave a blank line here.

File: contrib/nydusify/pkg/commiter/manager.go [hunk 16]
```diff
@@ -0,0 +1,130 @@
+package commiter
```
Comment (imeoer) [inline contrib/nydusify/pkg/commiter/manager.go:1]: Should add nydus developers copyright here.

File: contrib/nydusify/pkg/commiter/nsenter.go [hunk 17]
```diff
@@ -0,0 +1,185 @@
+// Ported from go-nsenter project, copyright The go-nsenter Authors.
+// https://github.com/Devatoria/go-nsenter
```
Comment (imeoer) [inline contrib/nydusify/pkg/commiter/nsenter.go:2]: Ditto, leave a blank line here.

File: smoke/tests/commit_test.go [hunk 18]
```diff
@@ -0,0 +1,123 @@
+package tests
```
Comment (imeoer) [inline smoke/tests/commit_test.go:1]: Ditto, add copyright here.

File: smoke/tests/commit_test.go [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+	tool.RunWithoutOutput(c.t, commitCmd)
+	// run committed container
+	runCommittedContainerCmd := fmt.Sprintf("sudo nerdctl --snapshotter nydus run  -d -t --insecure-registry --name=%s %s sh", committedContainerName, commmitedImage)
+	committedContainerID := strings.Trim(tool.RunWithOutput(runCommittedContainerCmd), "\n")
+
+	// check committed file content
+	checkCmd := fmt.Sprintf("sudo nerdctl cp %s:/root/commit container_commit", committedContainerID)
+	tool.RunWithoutOutput(c.t, checkCmd)
+
+	file, err := os.Open("./container_commit")
```
Comment (imeoer) [inline smoke/tests/commit_test.go:74]: We'd better put all temp files into the `ctx.Env.WorkDir` directory.

File: smoke/tests/commit_test.go [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+	if err != nil {
```
Comment (imeoer) [inline smoke/tests/commit_test.go:75]: Should use `require.NoError(t, err)`.

File: smoke/tests/commit_test.go [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+		c.t.Fatalf("open commit file failed: %s", err)
+	}
+	defer file.Close()
+	defer os.Remove(file.Name())
+
+	content, err := io.ReadAll(file)
+	if err != nil {
+		c.t.Fatalf("can not read commit file: %s", err)
+	}
+	fileContent := strings.Trim(string(content), "\n")
```
Comment (imeoer) [inline smoke/tests/commit_test.go:85]: Ditto, should use `require.NoError(t, err)`.

File: smoke/tests/commit_test.go [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+	committedContainerName := fmt.Sprintf("%s-committed", containerName)
+
+	commitCmd := fmt.Sprintf("sudo %s commit --container %s --target %s", ctx.Binary.Nydusify, containerID, commmitedImage)
+	tool.RunWithoutOutput(c.t, commitCmd)
+	// run committed container
+	runCommittedContainerCmd := fmt.Sprintf("sudo nerdctl --snapshotter nydus run  -d -t --insecure-registry --name=%s %s sh", committedContainerName, commmitedImage)
+	committedContainerID := strings.Trim(tool.RunWithOutput(runCommittedContainerCmd), "\n")
+
+	// check committed file content
+	checkCmd := fmt.Sprintf("sudo nerdctl cp %s:/root/commit container_commit", committedContainerID)
```
Comment (imeoer) [inline smoke/tests/commit_test.go:71]: May be implement a func like:

```
func checkFileContent(t *testing.T, containerName, path, content string) {
	nerdctlExec(t, containerName, fmt.Sprintf("stat %s && grep -Fxq '%s' %s", path, content, path))
}

func nerdctlExec(t *testing.T, containerName, cmd string) {
	tool.Run(t, fmt.Sprintf("nerdctl exec %s sh -c \"%s\"", containerName, cmd))
}
```

To check the file content consistency.

File: smoke/tests/commit_test.go [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+	}
+	defer file.Close()
+	defer os.Remove(file.Name())
+
+	content, err := io.ReadAll(file)
+	if err != nil {
+		c.t.Fatalf("can not read commit file: %s", err)
+	}
+	fileContent := strings.Trim(string(content), "\n")
+	if fileContent != "This is Nydus commit" {
```
Comment (imeoer) [inline smoke/tests/commit_test.go:86]: Ditto, `require.Equal`.

File: smoke/tests/commit_test.go [hunk 24]
```diff
@@ -0,0 +1,123 @@
...
+		}
+	}
+}
+
+func (c *CommitTestSuite) TestCommitAndCheck(ctx tool.Context, image, commmitedImage string) {
+
+	// run nydus contaienr
+	containerName := uuid.NewString()
+	runContainerCmd := fmt.Sprintf("sudo nerdctl --snapshotter nydus run  -d -t --insecure-registry --name=%s %s sh", containerName, image)
+	containerID := strings.Trim(tool.RunWithOutput(runContainerCmd), "\n")
```
Comment (imeoer) [inline smoke/tests/commit_test.go:54]: `defer tool.ClearContainer(c.t, image, "nydus", containerName)`

File: smoke/tests/commit_test.go [hunk 25]
```diff
@@ -0,0 +1,123 @@
...
+	defer os.Remove("./commit")
+
+	// commit container
+	committedContainerName := fmt.Sprintf("%s-committed", containerName)
+
+	commitCmd := fmt.Sprintf("sudo %s commit --container %s --target %s", ctx.Binary.Nydusify, containerID, commmitedImage)
+	tool.RunWithoutOutput(c.t, commitCmd)
+	// run committed container
+	runCommittedContainerCmd := fmt.Sprintf("sudo nerdctl --snapshotter nydus run  -d -t --insecure-registry --name=%s %s sh", committedContainerName, commmitedImage)
+	committedContainerID := strings.Trim(tool.RunWithOutput(runCommittedContainerCmd), "\n")
```
Comment (imeoer) [inline smoke/tests/commit_test.go:68]: `defer tool.ClearContainer(c.t, commmitedImage, "nydus", committedContainerName)`

**Comment by imeoer** [comment]
image

It seems the CI is broken.

**Comment by PerseidMeteor** [comment]
> image
> 
> It seems the CI broken.

It made me confused. Is there any idea to slove this problem?

**Review by imeoer — APPROVED** [review]

**Pull request merged**